### ISSUE TITLE
Rails 7 support?

### ISSUE THREAD

**Issue by mmhan** [gonzalo-bulnes/simple_token_authentication#385]
Given than rails 7 has released an alpha version, is rails 7 support coming to simple_token_authentication soon?

**Comment by mmhan** [comment]
I'm interested to contribute to this repo by resolving this issue. However, this will be the first time I will be opening a PR on github and I'm in need of some guidance on how to proceed.

So far, I have forked the repository, cloned it and have it run existing tests using `rake spec` with success.

As the next step, I generated new config for appraisal. Note: we need to use the main branch of devise for [rails 7 compatibility](https://github.com/heartcombo/devise/pull/5357). 

```
appraise 'rails_7_devise_4' do
  gem "actionmailer", "~> 7.0.0.alpha2"
  gem "actionpack", "~> 7.0.0.alpha2"
  gem "activerecord", "~> 7.0.0.alpha2"
  gem "devise", git: 'https://github.com/heartcombo/devise'
end
```

With that done, I found that mongoid is breaking on rails 7 too with the error. `uninitialized constant ActiveModel::Serializers::Xml` This is where I am stuck at because [mongoid's JIRA](https://jira.mongodb.org/projects/MONGOID/issues/MONGOID-3999?filter=allopenissues) does not have any related issue about rails 7.

Please let me know how best I can help to move this forward.

**Comment by amorimlucas** [comment]
We need this done too. Is there a maintainer you could contact directly @mmhan?

**Comment by OskarEichler** [comment]
@gonzalo-bulnes Us too 👍🏼

**Comment by OskarEichler** [comment]
@mmhan Maybe try again now that Rails 7 is officially released. Also found this:

https://stackoverflow.com/questions/43254528/nameerror-uninitialized-constant-activemodelserializersxml-when-declaring-a

Might need to add those as dependencies - not sure but worth a shot:

gem 'activemodel-serializers-xml'
gem 'active_model_serializers'

**Comment by gonzalo-bulnes** [comment]
Hi @mmhan,

Opening a PR is the right way to go! :heart: 
It has happened in the past that Mongoid support for new versions of Rails [took a bit of time to be realeased](https://github.com/gonzalo-bulnes/simple_token_authentication/pull/353), and we've worked around that (the linked commit was ther esoluton of the workaround).

Note that as long as Devise's support for Rails 7 is not released, we'll have to wait. This gem relies heavily on Devise, and relying on unstable versions of Devise is not an option. Don't let that prevent you from preparing your PR from Devise's `main` branch! Eventually, it will get released :slightly_smiling_face: 

With regards to @OskarEichler suggestion: I haven't looked at the situation in detail, but some parts of Rails are sometimes extracted to gems. That is usually documented in the Rails migration docs from one version to another.

**Comment by gonzalo-bulnes** [comment]
See also: https://github.com/gonzalo-bulnes/simple_token_authentication/issues/367

**Comment by gonzalo-bulnes** [comment]
You may already have seen them @mmhan, but the updates related to the two previous major Rails releases are a good reference to get an idea of what adding support for a new version of Rails looks like:

Add Rails 5 support: https://github.com/gonzalo-bulnes/simple_token_authentication/pull/262
Add Rails 6 support: https://github.com/gonzalo-bulnes/simple_token_authentication/pull/349

**Comment by OskarEichler** [comment]
@gonzalo-bulnes Thanks so much for looking into this! As far as I can tell the main branch of Devise is the stable branch - the older stable branches haven't been updated in years, and the main branch is very actively maintained. 

Maybe worth opening an issue and asking if the main branch can be considered stable and used. Will do that right now ;)

**Comment by gonzalo-bulnes** [comment]
Note @OskarEichler that even if the `main` Devise branch is actively maintained (I have all reasons to think it is!) doesn't make it more fit to be referenced as a dependency. Let me explain.

When any dependency is updated, it is possible that its changes are incompatible with the way you use them (depend on them). If that were to happen, any installation of your gem would stop working the way it is expected (which may or not be obvious, but _could_ be hard to detect).

Usually, with well-maintained gems, like Devise, the features that are dependable are documented, and their maintainers communicate when breaking changes happen by choosing carefully the new release number (see [semantic versioning](https://semver.org)).

If any gem doesn't reference its dependencies by release number... well, it cannot take advantage of that information. :woman_shrugging:  On the contrary, if a gem references its dependencies by release number, when a new release breaks important assumptions nothing happens, because the previous release remains in use until explicitly updated. :1st_place_medal: 

That is what I meant when I said "unstable" - if it was based on the `main` label, the _relationship_ between both gems would be unstable in that the assumptions on which it is built could change without notice. In other words Devise vX.Y.Z will always have the same feature it had on release day, but Devise `main` features change all the time by design. :slightly_smiling_face: 

**Summary**: Devise will eventually create a release that supports Rails 7, until then, we're not ready to support it either! Does this make sense?

**Comment by OskarEichler** [comment]
Hey @gonzalo-bulnes, thanks so much for the long explanation - makes total sense.

They did release version 4.8.1 with Rails 7 support on the main branch ([here](https://github.com/heartcombo/devise/releases))- why not just lock in dependancy at that version?

**Comment by gonzalo-bulnes** [comment]
Great @OskarEichler, that solves the Devise requirement. 4.8.1 is within the [existing version constraints for Devise][ref], which means that Devide 4.8.1 cam already be used with this gem, and there isn't anything special that need to be done for that.

On the testing side of things, now the appraisal that @mmhan got started (above) can be updated to reference Devise 4.8.1 ✅ ([Appraisals][app] allow to test specific combinations of dependencies together.)

Then will likely remain to follow up on Mongoid in the same way 🙂

  [ref]: https://github.com/gonzalo-bulnes/simple_token_authentication/blob/e9b09100eae032681b165194cef32d42395a9794/simple_token_authentication.gemspec#L21
  [app]: https://github.com/thoughtbot/appraisal

**Comment by mmhan** [comment]
I'm afraid things are a little too much at home right now with all the activities related to settling in after recently migrating from one country to another. I will get back to this once all the dust has settled. Feel free to take over from here if anyone wishes to. If not, I'll drop a comment when I can get around to it.

**Comment by gonzalo-bulnes** [comment]
No rush on my side! Thanks for following up @mmhan (and sharing your progress in the first place!)

It seems @OskarEichler is on a good track, and maybe this will incentive others to join the thread 🙂

**Comment by OskarEichler** [comment]
@gonzalo-bulnes Okay I tested all of this, thanks for the help here @mmhan 

The issue this is failing is because mongoid hasn't been updated to support Rails 7 yet, see here:
https://github.com/mongodb/mongoid/discussions/5107

When running Appraisal it actually resolves mongoid to an old beta version (v4.0.0beta) because that probably didn't have a gemspec. It's not compatible and once forcing the latest version of Mongoid by specifying it in the Appraisal file the bundler fails because it is not yet compatible:

`appraise 'rails_7_devise_4' do
  gem "actionmailer", "~> 7"
  gem "actionpack", "~> 7"
  gem "activerecord", "~> 7"
  gem "devise", git: 'https://github.com/heartcombo/devise'
  gem "activemodel-serializers-xml"
  gem "active_model_serializers"
  gem "mongoid", "7.3.3"
end`

Will need to wait until Mongoid is updated...

**Comment by OskarEichler** [comment]
@gonzalo-bulnes Actually it seems like mongoid is only a dependancy for development:

`s.add_development_dependency 'mongoid', ">= 3.1.0", "< 8"`

Can this potentially be ignored or why do you need it in development mode?

**Comment by amorimlucas** [comment]
Looks like Mongoid was updated to support Rails 7 a few days ago. Was that the only holding this up?

**Comment by OskarEichler** [comment]
@amorimlucas Just updated Mongoid and looks like everything is working now indeed!

@gonzalo-bulnes Made a PR here for you to review and merge: #391

**Comment by amorimlucas** [comment]
@gonzalo-bulnes Any progress on this? Is there someone else that could make this merge? Thx

**Comment by GregSmith92** [comment]
Just wanted to check in and see #391 is any closer to being merged? 🙏 🙏 🙏

**Comment by aried3r** [comment]
Using #391 as a base, I ran the tests using `appraisal` for Rails 5, 6 and 7 using Ruby 2.5, 2.7, 3.0 and 3.1 after relaxing the development dependency in the `.gemspec` file (see https://github.com/gonzalo-bulnes/simple_token_authentication/pull/391/files#r874883121).

Ruby 2.5 (except with Rails 7, which only support 2.7+) and 2.7 complete successfully, but Ruby 3.0 and 3.1 have three test failures on all Rails versions. All three are almost identical failures and while I haven't looked into them yet, I believe it's because of Ruby 3.0+'s kwarg handling, but I could be wrong.

Also I don't think these are related to Rails 7 support so are not necessarily blocking. 

The test failures looks like this:

```
  1) Simple Token Authentication :sign_in_token option can be modified from an initializer file
     Failure/Error: controller.send(:sign_in, record, *args)

       #<#<Class:0x0000000113d93050>:0x0000000113d81c60> received :sign_in with unexpected arguments
         expected: (#<Double (anonymous)>, {:store=>"updated value"})
              got: (#<Double (anonymous)>, {:store=>"updated value"})
```

@gonzalo-bulnes, would you be open to a little cleanup of supported Ruby and Ruby on Rails versions? Rails 5.2 hits EOL [in two weeks](https://endoflife.date/rails), Ruby 2.6 hit EOL [a month ago](https://endoflife.date/ruby), but personally I think maintaining EOL Rubies like 2.5 is easier than older Rails versions. `devise` version 4 was [released in 2016](https://github.com/heartcombo/devise/blob/main/CHANGELOG.md#400---2016-04-18).

Since old versions of the gem would continue to work for users of older versions, I'd like to propose the following cleanup:

* Test Ruby 2.7, 3.0, 3.1 (I'd go as low as 2.5 if necessary)
* Test Rails 6.0, 6.1, 7.0 (5.2 only if absolutely necessary)
* Test devise 4.x (dropping support for 3.x)

What do you think?

PS: I haven't taken a closer look at mongoid support and if `bundle exec appraisal rake` tests it or if I have to do something special. I went with what the CI setup did.

**Comment by pglombardo** [comment]
As an outside observer looking for Rails 7 support, that sounds like a good plan @aried3r.  😄 

Any update on Rails 7 support?  I'd like to upgrade [the opensource project I maintain](https://github.com/pglombardo/PasswordPusher) but am blocked by this gem (which is a great and very helpful gem).

Willing to jump in if needed.

**Comment by gonzalo-bulnes** [comment]
Hi everyone! I'm catching up on the thread.

> Actually it seems like mongoid is only a dependancy for development [...]
Can this potentially be ignored or why do you need it in development mode?

Mongoid is a "peer dependency". It made little sense to me to require it if you don't use it, but since support for Mongoid was added, I wouldn't remove it without careful consideration (and a major versiom change).

**Comment by gonzalo-bulnes** [comment]
@aried3r You're certainly making a good argument for scoping the supported Rails version range. In my experience, Rails support tends indeed to require more compromises than supporting older Rubies. ([example])

I have tried to keep support until now to avoid breaking backwards compatibility. More by principle than for any practical reason, so it may be time to do a pragmatic pruning and make a `2.0.0` release.

I'll get to the test failures in a separate comment.



  [example]: https://github.com/gonzalo-bulnes/simple_token_authentication/blob/v1.17.0/lib/simple_token_authentication.rb

**Comment by gonzalo-bulnes** [comment]
@aried3r Regarding the test failures, the output is certainly not very helpful! Your hypothesis makes a lot of sense to me. I'm not sure how to confirm that, but it seems like getting a replacement for Travis CI would be a good start. (I've seem your suggestion of GitHub Actions in #391)

**Comment by gonzalo-bulnes** [comment]
Adding to my comment on @aried3r proposal for support pruning: dropping support for `devise < 4` seems reasonable to me in a `2.0.0` release. Devise is a security-critical dependency, and as much as I don't think it's helpful to over-constrain requirments, a little nudge to upgrade Devise seems reasonable to me.

**Comment by gonzalo-bulnes** [comment]
To everyone in the thread, thank you. I know my involvment has been long to come, and I appreciate how you've stepped in to research, discuss, write code for, test and provide feedback on #391.

Maintaining whatever project is a significant amount of work in the long run, and I think it's fair to say that my earlier attempts to get this kind of dynamics going were not successful. I'm very glad to see this (especially because it's been somewhat unexpected!), and I want to acknowledge @mmhan @OskarEichler @aried3r @pglombardo and everybody else involved for your work on this. ❤️

In the meantime, Travis CI changed their policies, which is a bit of a set back. I'd love to get some alternative going so that we can release either `1.18.0` or `2.0.0` in good conditions. I'll do my research on GitHub Actions per @aried3r suggestion (I haven't used them yet). In case it's not obvious: any input or code is appreciated, as you might have noticed that I've had a few competing priorities 😐 but I'll do my best to support the collective effort to get the release out!

**Comment by pglombardo** [comment]
> Travis CI changed their policies, which is a bit of a set back. I'd love to get some alternative going so that we can release either 1.18.0 or 2.0.0 in good conditions.

Feel free to copy [.github/workflows/ruby.yml file](https://github.com/pglombardo/simple_token_authentication/blob/9ca489a7ba84a1ac10436805c5f3e060054e8eb6/.github/workflows/ruby.yml) to enable Github Actions.  Also suggested in [#391](https://github.com/gonzalo-bulnes/simple_token_authentication/pull/391#issuecomment-1309932821).

**Comment by letz** [comment]
what's the state of this? Does anyone need help to push this forward? I don't mind to help

**Comment by gonzalo-bulnes** [comment]
Closed via #401
